### PR TITLE
docs: Provide Azure Application Insights scaler

### PR DESCRIPTION
**Pull request by markrzasa**
Signed-off-by: Mark Rzasa <mark.rzasa@gmail.com>

This is the documentation for an Azure Application Insights scaler. This is related to this issue:

https://github.com/kedacore/keda/issues/1965

This pull request is marked as a draft because an app insights scaler may not be necessary to resolve issue 1965. Alternatively, the Azure Log Analytics scaler could be used to read app insights metrics from the AppMetrics table.

### Checklist

- [x] Commits are signed with Developer Certificate of Origin (DCO)

Fixes #1965

### CODE REVIEW

File: content/docs/2.6/scalers/azure-app-insights.md [hunk 1]
```diff
@@ -0,0 +1,120 @@
++++
+title = "Azure Application Insights"
+layout = "scaler"
+availability = "v1.3+"
```
Comment (tomkerkhove) [inline content/docs/2.6/scalers/azure-app-insights.md:4]:
```suggestion
availability = "v2.6+"
```

File: content/docs/2.6/scalers/azure-app-insights.md [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+
+```yaml
+triggers:
+- type: azure-app-insights
+  metadata:
+    metricAggregationTimespan: "0:1"
+    metricAggregationType: avg
+    metricFilter: cloud/roleName eq 'role_name'
+    metricName: "metric_name"
+    metricNamespace: "customMetrics"
```
Comment (tomkerkhove) [inline content/docs/2.6/scalers/azure-app-insights.md:22]: Do we need the metric namespace here?

File: content/docs/2.6/scalers/azure-app-insights.md [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+    targetValue: "1"
+    activeDirectoryClientIdFromEnv: CLIENT_ID_ENV_NAME # Optional, can use TriggerAuthentication as well
+    activeDirectoryClientPasswordFromEnv: CLIENT_PASSWORD_ENV_NAME # Optional, can use TriggerAuthentication as well
+    appIdFromEnv: APP_ID # Optional, can use TriggerAuthentication as well
```
Comment (tomkerkhove) [inline content/docs/2.6/scalers/azure-app-insights.md:26]:
```suggestion
    applicationInsightsNameFromEnv: APP_ID # Optional, can use TriggerAuthentication as well
```

Do you mean the name or ID? If you mean the ID, is that the ARM URI or?

File: content/docs/2.6/scalers/azure-app-insights.md [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+    tenantIdFromEnv: TENANT_ID` # Optional, can use TriggerAuthentication as well
+```
+
+This scaler is backed by the Azure Application Instance REST API. Please see [this](https://dev.applicationinsights.io/reference) page
+for further details.
+
+**Parameter list:**
```
Comment (tomkerkhove) [inline content/docs/2.6/scalers/azure-app-insights.md:33]: Can you align with our guidelines please? https://github.com/kedacore/keda-docs#writing-documentation-for-a-scaler

**Review by tomkerkhove — COMMENTED** [review]
Did a first skim, thanks a ton!

File: content/docs/2.6/scalers/azure-app-insights.md (hunk 2 above)
Comment (markrzasa) [reply content/docs/2.6/scalers/azure-app-insights.md]: I don't think so as it doesn't match the API. I've removed the parameter and updated the examples.

File: content/docs/2.6/scalers/azure-app-insights.md (hunk 3 above)
Comment (markrzasa) [reply content/docs/2.6/scalers/azure-app-insights.md:25]: I meant the ID or GUID. I tried to improve the description to make this more clear. Let me know if it helps.

File: content/docs/2.6/scalers/azure-app-insights.md (hunk 4 above)
Comment (markrzasa) [reply content/docs/2.6/scalers/azure-app-insights.md:32]: Hopefully, this is better. Please let me know if the doc is still not aligned.

**Comment by markrzasa** [comment]
> Did a first skim, thanks a ton!

No problem at all! Thanks for the comments.

File: content/docs/2.6/scalers/azure-app-insights.md (hunk 3 above)
Comment (tomkerkhove) [reply content/docs/2.6/scalers/azure-app-insights.md:25]: Thanks!

In that case I would rename the field to this:
```suggestion
    applicationInsightsIdFromEnv: APP_ID # Optional, can use TriggerAuthentication as well
```

Because app ID is very confusing in the combination with AD auth in my opinion

File: content/docs/2.6/scalers/azure-app-insights.md (hunk 4 above)
Comment (tomkerkhove) [reply content/docs/2.6/scalers/azure-app-insights.md:32]: Thank you!

File: content/docs/2.6/scalers/azure-app-insights.md (hunk 2 above)
Comment (tomkerkhove) [reply content/docs/2.6/scalers/azure-app-insights.md]: Actually I think my initial comment was wrong, sorry! Let me rephrase that: Do we need the `metric` prefix? Can't we just use `aggregationType`, `filter`, etc?

It would be good to keep in the `namespace` btw so that people can change it if they need to, but we can use `customMetrics` as a default.

File: content/docs/2.6/scalers/azure-app-insights.md [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+This specification describes the `azure-app-insights` trigger that scales based on an Azure Application Insights metric.
+
+```yaml
+triggers:
+- type: azure-app-insights
+  metadata:
+    metricAggregationTimespan: "0:1"
+    metricAggregationType: avg
+    metricFilter: cloud/roleName eq 'role_name'
+    metricName: "customMetrics/example-metric"
```
Comment (zroubalik) [inline content/docs/2.6/scalers/azure-app-insights.md:21]: Is this a `metricName` for KEDA internals? If so, we shouldn't requrie users to specify this. We should generate one, take a look on the last few scalers how they done this.

Comment (markrzasa) [reply content/docs/2.6/scalers/azure-app-insights.md:21]: This is the metric to retrieve from app insights. Some example are availabilityResults/availabilityPercentage, browserTimings/networkDuration, and customMetrics/customMetricForSomeService). The REST API described [here](https://dev.applicationinsights.io/reference) uses the term metricId. Maybe that's better?

Comment (zroubalik) [reply content/docs/2.6/scalers/azure-app-insights.md:21]: Gotcha, in this case I think that `metricId` is better, so we are aligned with the service and avoid confusion with KEDA's `metricName`

Comment (markrzasa) [reply content/docs/2.6/scalers/azure-app-insights.md:21]: Sounds good. I'll make that change later today.

File: content/docs/2.6/scalers/azure-app-insights.md (hunk 2 above)
Comment (markrzasa) [reply content/docs/2.6/scalers/azure-app-insights.md]: Some of this discussion is related to the discussion below around the metricName parameter. The REST API documented [here](https://dev.applicationinsights.io/reference) refers to the metric to be retrieved as the metricId. Some examples are:
- availabilityResults/availabilityPercentage
- browserTimings/networkDuration
- customMetrics/customMetricForSomeService

The suggestion below is to change metricName to metricId to be consistent with the API. If that's ok, that leaves us with the metricId, aggregationType, and filter fields. Those fields will allow us to do things like aggregate a metric using a given dimension.

Comment (tomkerkhove) [reply content/docs/2.6/scalers/azure-app-insights.md]: Sounds good to me!

File: content/docs/2.6/scalers/azure-app-insights.md (hunk 5 above)
Comment (tomkerkhove) [reply content/docs/2.6/scalers/azure-app-insights.md:21]: Sounds good to me!

**Comment by tomkerkhove** [comment]
Super excited about this one, scale based on custom metrics 💘

File: content/docs/2.6/scalers/azure-app-insights.md (hunk 2 above)
Comment (markrzasa) [reply content/docs/2.6/scalers/azure-app-insights.md]: Great! Thanks. I'll get those changes in later today.

File: content/docs/2.6/scalers/azure-app-insights.md (hunk 5 above)
Comment (markrzasa) [reply content/docs/2.6/scalers/azure-app-insights.md]: metricName has been renamed to metricId

**Comment by markrzasa** [comment]
I think I've covered all of the comments. Please let me know if anything else needs to be changed. Thanks.

File: content/docs/2.6/scalers/azure-app-insights.md [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+
+### Authentication Parameters
+
+You can use the `TriggerAuthentication` CRD to configure the authentication by providing a set of Azure Active Directory credentials or by using pod identity.
+
+**Credential based authentication:**
+
+- `activeDirectoryClientId` - Id of the Active Directory application which requires at least `Monitoring Reader` permissions.
+- `activeDirectoryClientPassword` - Password of the Active Directory application.
+
```
Comment (tomkerkhove) [inline content/docs/2.6/scalers/azure-app-insights.md:60]: `applicationInsightsId` seems to be missing here

File: content/docs/2.6/scalers/azure-app-insights.md [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+for further details.
+
+**Parameter list:**
+
+- `tenantId` - Id of the tenant that contains the Azure resource. This is used for authentication.
+- `metricId` - The name of the Application Insights metric to query. Run `az monitor app-insights metrics get-metadata` to see a list of available metrics.
+- `targetValue` - Target value to trigger scaling actions.
+- `metricAggregationType` - Aggregation method of the Azure Application Insights metric. The aggregation methods vary from metric to metric. The `az monitor app-insights metrics get-metadata` command can be used to determine which methods apply to a given metric. (Some common aggregation methods are `avg`, `count`, `sum`, `min`, and `max`)
+- `metricAggregationInterval` - Collection time of the metric in format `"hh:mm"`.
+- `applicationInsightsId` - Id of the Application Insights instance to query. This is a GUID that can be retrieved from the Application Insight's `API Access` blade in the Azure Portal.
```
[the remark this answers is not in the document]
Comment (markrzasa) [reply content/docs/2.6/scalers/azure-app-insights.md:39]: I was following the pattern found in some of the other scalers like the log analytics scaler. These scalers allow parameters to be read from an environment variable or an auth trigger. I'd be happy to remove the *FromEnv parameters as that would simplify things.

[the remark this answers is not in the document]
Comment (markrzasa) [reply content/docs/2.6/scalers/azure-app-insights.md:39]: Right. Good point. This isn't a secret. I'll move this out of the trigger and out of the *FromEnv fields.

File: content/docs/2.6/scalers/azure-app-insights.md [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+- `tenantId` - Id of the tenant containing the Application Insights instance.
+- `activeDirectoryClientId` - Id of the Active Directory client. The client must have `Monitoring Reader` permissions for the Application Insights instance.
+- `activeDirectoryClientPassword` - Password of the Active Directory client password.
```
[the remark this answers is not in the document]
Comment (markrzasa) [reply content/docs/2.6/scalers/azure-app-insights.md:42]: For now, yes. Is this something that should be added now or in a subsequent PR?

File: content/docs/2.6/scalers/azure-app-insights.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (tomkerkhove) [reply content/docs/2.6/scalers/azure-app-insights.md:39]: Let's do both - Let's keep it for those who want to have flexibility, but add declared field as well. Thoughts?

File: content/docs/2.6/scalers/azure-app-insights.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (tomkerkhove) [reply content/docs/2.6/scalers/azure-app-insights.md:42]: Follow-up PR based on end-user request is totally fine!

[the remark this answers is not in the document]
Comment (markrzasa) [reply content/docs/2.6/scalers/azure-app-insights.md:42]: It looks like the code already supports this.

File: content/docs/2.6/scalers/azure-app-insights.md [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+    tenantIdFromEnv: TENANT_ID` # Optional, can use TriggerAuthentication as well
+```
+
+This scaler is backed by the Azure Application Instance REST API. Please see [this](https://docs.microsoft.com/en-us/rest/api/application-insights/metrics/get) page
+for further details.
+
+**Parameter list:**
+
+- `tenantId` - Id of the tenant that contains the Azure resource. This is used for authentication.
+- `metricId` - The name of the Application Insights metric to query. Run `az monitor app-insights metrics get-metadata` to see a list of available metrics.
```
Comment (arschles) [inline content/docs/2.6/scalers/azure-app-insights.md:35]: Just an addition to link out to the Azure CLI doc, in case someone doesn't know or forgot how to use it.

```suggestion
- `metricId` - The name of the Application Insights metric to query. Use the [Azure Command Line Interface](https://docs.microsoft.com/en-us/cli/azure/install-azure-cli) to run `az monitor app-insights metrics get-metadata` to see a list of available metrics.
```

File: content/docs/2.6/scalers/azure-app-insights.md [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+- `metricAggregationInterval` - Collection time of the metric in format `"hh:mm"`.
+- `applicationInsightsId` - Id of the Application Insights instance to query. This is a GUID that can be retrieved from the Application Insight's `API Access` blade in the Azure Portal.
+- `tenantId` - Id of the tenant containing the Application Insights instance.
+- `activeDirectoryClientId` - Id of the Active Directory client. The client must have `Monitoring Reader` permissions for the Application Insights instance.
+- `activeDirectoryClientPassword` - Password of the Active Directory client password.
+- `metricFilter` - Further specify the metrics query using a filter. For example `cloud/roleName eq 'example`. (Optional)
+
+Some parameters can be provided using environmental variables, instead of setting them directly in metadata. Here is a list of parameters you can use to retrieve values from environment variables:
```
Comment (arschles) [inline content/docs/2.6/scalers/azure-app-insights.md:45]: small fix:

```suggestion
Some parameters can be provided using environment variables, instead of setting them directly in metadata. Here is a list of parameters you can use to retrieve values from environment variables:
```

Also, is it possible to set those environment variables from the helm chart? Or from a `ScaledObject`. Regardless, it might be good to briefly explain how to do it, or link out to somewhere that does.

File: content/docs/2.6/scalers/azure-app-insights.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (markrzasa) [reply content/docs/2.6/scalers/azure-app-insights.md:39]: OK. Sounds good. The code handles this case already.

File: content/docs/2.6/scalers/azure-app-insights.md (hunk 6 above)
Comment (markrzasa) [reply content/docs/2.6/scalers/azure-app-insights.md:62]: Thanks for catching that. Should be fixed now.

File: content/docs/2.6/scalers/azure-app-insights.md (hunk 9 above)
Comment (markrzasa) [reply content/docs/2.6/scalers/azure-app-insights.md]: Good idea. Thanks. Fixed.

File: content/docs/2.6/scalers/azure-app-insights.md (hunk 10 above)
Comment (markrzasa) [reply content/docs/2.6/scalers/azure-app-insights.md]: I've added an example at the bottom of the page. Let me know if that looks ok.

File: content/docs/2.6/scalers/azure-app-insights.md [hunk 11]
```diff
@@ -0,0 +1,197 @@
...
+    applicationInsightsIdFromEnv: APP_ID # Optional, can use TriggerAuthentication as well
+    tenantIdFromEnv: TENANT_ID` # Optional, can use TriggerAuthentication as well
+```
+
+This scaler is backed by the Azure Application Instance REST API. Please see [this](https://docs.microsoft.com/en-us/rest/api/application-insights/metrics/get) page
+for further details.
+
+**Parameter list:**
+
+- `tenantId` - Id of the tenant that contains the Azure resource. This is used for authentication.
```
Comment (tomkerkhove) [inline content/docs/2.6/scalers/azure-app-insights.md:34]: This one is in the list multiple times

File: content/docs/2.6/scalers/azure-app-insights.md [hunk 12]
```diff
@@ -0,0 +1,197 @@
...
+    activeDirectoryClientIdFromEnv: CLIENT_ID_ENV_NAME # Optional, can use TriggerAuthentication as well
+    activeDirectoryClientPasswordFromEnv: CLIENT_PASSWORD_ENV_NAME # Optional, can use TriggerAuthentication as well
+    applicationInsightsIdFromEnv: APP_ID # Optional, can use TriggerAuthentication as well
+    tenantIdFromEnv: TENANT_ID` # Optional, can use TriggerAuthentication as well
+```
+
+This scaler is backed by the Azure Application Instance REST API. Please see [this](https://docs.microsoft.com/en-us/rest/api/application-insights/metrics/get) page
+for further details.
+
+**Parameter list:**
```
Comment (tomkerkhove) [inline content/docs/2.6/scalers/azure-app-insights.md:32]: Sanity check - Are we listing all parameters here?

File: content/docs/2.6/scalers/azure-app-insights.md [hunk 13]
```diff
@@ -0,0 +1,197 @@
...
++++
+
+### Trigger Specification
+
+This specification describes the `azure-app-insights` trigger that scales based on an Azure Application Insights metric.
+
+```yaml
+triggers:
+- type: azure-app-insights
+  metadata:
```
Comment (tomkerkhove) [inline content/docs/2.6/scalers/azure-app-insights.md:17]: Sanity check - Are we showcasing all parameters here?

File: content/docs/2.6/scalers/azure-app-insights.md (hunk 11 above)
Comment (markrzasa) [reply content/docs/2.6/scalers/azure-app-insights.md:34]: Fixed. Thanks for catching that.

File: content/docs/2.6/scalers/azure-app-insights.md (hunk 13 above)
Comment (markrzasa) [reply content/docs/2.6/scalers/azure-app-insights.md:17]: I believe so.

File: content/docs/2.6/scalers/azure-app-insights.md (hunk 12 above)
Comment (markrzasa) [reply content/docs/2.6/scalers/azure-app-insights.md:32]: A note has been added for pod identity. That should cover everything.

**Review by tomkerkhove — APPROVED** [review]
